chess-AI/experimental-work/mcts.py

```diff
@@ -1,73 +1,161 @@
+                if uci_move in self.q_values[fen_string]:
+                    q_value = self.q_values[fen_string][uci_move]
+                else:
+                    self.q_values[fen_string][uci_move] = 0
+                    q_value = 0
+            else:
+                self.q_values[fen_string]={}
+                self.q_values[fen_string][uci_move] = 0
+                q_value = 0
+
+
+            # Check to see if state has been added to N
+            if fen_string in self.n_values:
+                if uci_move in self.n_values[fen_string]:
+                    n_value = self.n_values[fen_string][uci_move]
+                else:
+                    self.n_values[fen_string][uci_move] = 1
+                    n_value = float("inf")
+            else:
+                self.n_values[fen_string]={}
+                self.n_values[fen_string][uci_move] = 1
+                n_value = float("inf")
+
+            """
+            NOTE: still unsure about the n values
+            not sure if I should set the n values to infinite if a node
+            has not been visited
+            Also do not know what N should be initial
+            """
+
+            # Calculate U based on the UCB formula
+            u_value = q_value + (self.exploration * self.p_values[fen_string][uci_move])
+            u_value = u_value * sqrt(sum(self.n_values[fen_string].values())/n_value)
+
+            """
+            NOTE: Look at different formulas for calculating U
+            There could be better formula for evaluating an action
+            """
+
+            # Want to choose the move with the highest u value
+            if u_value > best_u:
+                best_u = u_value
+                best_move = move # move with best u value
+
+        return best_move
 
-        for a in game.getvalidActions(s): # finds the valid actions from current state
 
-            u = Q[s][a] + c_puct*P[s][a]*sqrt(sum(N[s]))/(1+N[s][a]) # formula for calculating the action to take
+    # Function to calculate Q and N based on the value and the move
+    def update_qn(self,fen_string,move,value):
 
-            if u>max_u:
-                max_u = u  # assigns best u
-                best_a = a  # assigns best a
-        
-            a = best_a
-            sp = game.nextState(s,a) # want to take the the best action
-            v = search(sp,game,nnet) # gets the value of the the action
+        move = move.uci()
 
-            Q[s][a] = (N[s][a] * Q[s][a] +v)/N[s][a]+1) # calculates Q
-            N[s][a] += 1  # adds 1 to represent the node was visited
+        # Function to calculate Q
+        self.q_values[fen_string][move] = self.n_values[fen_string][move] * self.q_values[fen_string][move] + value
+        self.q_values[fen_string][move] = self.q_values[fen_string][move] / (self.n_values[fen_string][move] + 1)
 
-            return -v
+        """
+        NOTE: Can look into different formulas for this as well
+        """
```

Comment:
Should be `# TODO`

Reply:
Also, typically comments refer to the code below the comment. If this refers to the q_values computation, please move it above that code (and same elsewhere)